aggregator_core/src/datastore.rs

```diff
@@ -1470,81 +1470,48 @@ impl<C: Clock> Transaction<'_, C> {
             return Ok(());
         }
 
-        // Slow path: no rows were affected, meaning a row with the new report ID already existed
-        // and we hit the query's ON CONFLICT DO NOTHING clause. We need to check whether the new
-        // report matches the existing one.
-        let existing_report = {
-            // We intentionally don't use `get_client_report` because it omits expired reports. It
-            // is possible that we have conflicted with an expired report that hasn't been fully
-            // GC'd yet.
-            let stmt = self
-                .prepare_cached(
-                    "SELECT
-                        client_reports.client_timestamp,
-                        client_reports.extensions,
-                        client_reports.public_share,
-                        client_reports.leader_input_share,
-                        client_reports.helper_encrypted_input_share
-                    FROM client_reports
-                    JOIN tasks ON tasks.id = client_reports.task_id
-                    WHERE tasks.task_id = $1
-                      AND client_reports.report_id = $2",
-                )
-                .await?;
+        // No rows were affected, meaning a row with the report ID already existed and we hit the
+        // query's ON CONFLICT DO NOTHING clause.
+        Err(Error::MutationTargetAlreadyExists)
+    }
 
-            self.query_opt(
+    // scrub_client_report removes the client report itself from the datastore, retaining only a
+    // small amount of metadata required to perform duplicate-report detection & garbage collection.
+    //
+    // This method is intended for use by aggregators acting in the Leader role. Scrubbed reports
+    // can no longer be read, so this method should only be called once all aggregations over the
+    // report have stepped past their START state.
+    pub async fn scrub_client_report(
+        &self,
+        task_id: &TaskId,
+        report_id: &ReportId,
+    ) -> Result<(), Error> {
+        let stmt = self
+            .prepare_cached(
+                "UPDATE client_reports SET
+                extensions = NULL,
+                public_share = NULL,
+                leader_input_share = NULL,
+                helper_encrypted_input_share = NULL,
+                updated_at = $1,
+                updated_by = $2
+            FROM tasks
+            WHERE tasks.id = client_reports.task_id
+              AND tasks.task_id = $3 AND client_reports.report_id = $4",
```

Comment:
Does pg actually immediately reclaim the space gained by setting these NULL? Does it require a vacuum?

I will experiment/research, but curious if you know the answer off hand.

Reply:
Since Postgres uses MVCC, this UPDATE results in new tuples being inserted. The existing tuple becomes a dead tuple once enough transactions have finished, and the space occupied by the dead tuple gets freed by autovacuum workers. (FWIW, DELETE'd tuples need help from autovacuum workers to complete their lifecycle as well, the only difference is that there's no new tuple inserted in the table)

The story gets more complicated if rows are big enough that they don't fit in a page, as report shares may get stored via the TOAST system. In that case, the row is represented by both a tuple in the main table, and one or more tuples in TOAST tables. Autovacuuming applies to the TOAST table as well.

Reply:
David's explanation matches my understanding. This behavior is why I'm nervous about increasing our rate of writes -- it's hard to reason about how Postgres' autovacuum will handle this. (That is, we may be saving a lot of space, but incurring a bunch of extra work for vacuum.)